lib/sender.js

```diff
@@ -81,14 +86,24 @@ class Sender {
 
     if (!options.mask) return [target, data];
 
-    randomFillSync(mask, 0, 4);
+    let mask = _mask;
+
+    if (options.generateMask) {
+      mask = options.generateMask();
+    } else {
+      randomFillSync(_mask, 0, 4);
+    }
 
     target[1] |= 0x80;
     target[offset - 4] = mask[0];
     target[offset - 3] = mask[1];
     target[offset - 2] = mask[2];
     target[offset - 1] = mask[3];
 
+    if ((mask[0] | mask[1] | mask[2] | mask[3]) === 0) {
```

Comment:
I think this optimization does not make sense. It is actually counterproductive in the common case. If a random mask is used, then the probability of this event is 1 / 2**32. The same applies for the same optimization on `lib/receiver.js`.

Reply:
I don’t think it hurts and allows for hot path optimization for clients that know the hotel path “trick”.

Reply:
I benchmarked this line `(mask[0] | mask[1] | mask[2] | mask[3]) === 0` and it came out to less than 1 nanosecond, which is basically zero length of time. I benchmarked the time to mask a 10kb string with `applyMask(data, mask, target, offset, data.length)` and it came to 1.4 milliseconds, 3 million times larger. Masking 100 10kb strings would take 140ms, checking for mask 100 times would 50 nano seconds. The larger the string, the larger the mask time, while checking mask stays constant <1 nanosecond. That being said, if you want the best performance you don't need to check the mask every ws.send, you only need to check it on initial websocket setting, but currently the PR has it a function that must be generated every ws.send. Though 1 nanosecond may be worth the effort if you want the simplified generateMask function option

Reply:
If you want best performance you should accept a Buffer or Function for `generateMask` option

Reply:
Probably the biggest performance issue is repeatedly creating new Buffer objects in `generateMask`, this should be re-coded to allow accepting a Buffer option, and the Function method should modify the existing Buffer instead of generate new one as suggested here https://github.com/websockets/ws/pull/1989#issuecomment-997392448

Reply:
I initially called the option `fillMaskBuffer`, and made it take the mask buffer like `crypto.randomFillSync()`. Then I changed my mind and went for `generateMask`. The reason is that it is simpler to use. The user does not have to remember to fill the buffer. They are not forced to do that but we should then document that the last one is used even if the `WebSocket` instance is different. Not returning a buffer results in a thrown error.

Finally, `generateMask` does not force the user to return a new `Buffer` instance every time. The user can return the same instance every time:

```js
const mask = Buffer.alloc(4);

const ws = new Websocket(url, {
  generateMask() {
    mask.fill(0);
    return mask;
  }
});
```

Reply:
Isn't this sufficient? I don't think `ws` will mutate the buffer, right?

```js
const mask = Buffer.alloc(4);

const ws = new Websocket(url, {
  generateMask() {
    return mask;
  }
});
```

Reply:
Correct.

Reply:
@e3dio 

> I benchmarked this line `(mask[0] | mask[1] | mask[2] | mask[3]) === 0` and it came out to less than 1 nanosecond

How can you get that precision? Two sequential `process.hrtime.bigint()` with nothing between take ~0.1 milliseconds / ~100000 nanoseconds on my machine.

Reply:
When I do 1 iteration I get 2000 nanosec, doing a loop benchmark gets you to .4 nanosec, I assume JS engine is going to optimize that code for a hot path (hot vs slow execution), not sure how you are getting 100,000 nano sec for 1 iteration though that seems off, are you sure you have the numbers right? 0.1 milliseconds = 100,000,000 nanoseconds so I think you have wrong number
```javascript
const mask = Buffer.from([1,2,3,4]);
const start = process.hrtime.bigint();
(mask[0] | mask[1] | mask[2] | mask[3]) === 0;
console.log(process.hrtime.bigint() - start)
```

Reply:
ok you got 100,000 nanosec, did you try the above code?

Reply:
I tested the `applyMask` function again, looks like in my 1st post I gave the 1 iteration benchmark, so to compare correctly against mask check I should have done the loop benchmark also, I am getting 14,000 nanosec for `applyMask`, so that function is not that bad, only 35,000 times worse than mask check, not 2 million times worse, I should update that post

Reply:
But probably the most impactful thing is skipping the crypto generation which I haven't even tried benchmark yet

Reply:
Why don't you make an uint32array from the memory and just check if (a[0])

Reply:
I tried `buf.readInt32BE()===0` it gives 1.81 nanosec vs the existing is .45 nanosec. .45 is hard to beat, that is including the integer iteration which probably takes up a chunk. 8700k CPU is 4.70GHz turbo so that's almost 5 cpu cycles per nanosecond, .45ns takes up 2 cpu cycles, hard to beat that

Reply:
Buffer.readUInt32BE is implemented as a _function_ by Node.js guys. Uint32Array is an _intrinsic_ to the very V8 engine. It should be just a jz. But, in any case, this is just nerd talk as we are literally discussing an if-statement involving 4 bytes. None of this matters in the given context.

Reply:
I was going to say this is the most insignificant thing possible to discuss improving on .45ns but you were the one that brought it up so I gave it a try

Reply:
Ah, my idea was more about _simplifying the code_. You don't need to OR together 4 individual bytes if you simply have an uint32array. Passing an uint32array around is also simpler to fill, since you can just pick a random integer between 0 and uint32_max (which is exactly what rand() returns). Point being, an uint32 is simpler dealing with IMO. Never mind.

Reply:
uint32 is also simpler dealing with if you want to have a mask that just increments by 1 every time. IMO uint32 is the perfect data type for the mask.

Reply:
Ok sorry I see what you are saying, pass around Uint32Array's instead of Buffers for the mask. You were throwing me off randomly commenting in this thread, he was asking me about benchmarking what amounts to an equals statement which is non-productive, but using Uint32Array is a valid point
```javascript
uint32 = new Uint32Array(1) // new mask default value = 0

{ generateMask: uint32 => require('crypto').randomFillSync(uint32) }
{ generateMask: uint32 => uint32[0] = Math.random()*2**32 }
{ generateMask: uint32 => uint32[0]++ }

const skipMasking = uint32[0] === 0;
```
They are easier to work with than Buffer because the mask is 32 bit unsigned int. Maybe only reason to use Buffer is more people are familiar with Buffer? You could do either one, Uint32Array makes more sense, you just have to know how it works. Buffer is easier to say than Uint32Array. Ideally Uint32Array but Buffer is ok

Reply:
> Maybe only reason to use Buffer is more people are familiar with Buffer?

Yes, and because `bufferUtil.mask()` was written to take a buffer at the time https://github.com/websockets/bufferutil/blob/cd48704aa8e87c19d7497812d466fd6fb6a28c09/src/bufferutil.cc#L89-L90, when typed array were not supported, and that did not change over the time. We could make it take a `Uint32Array` but I do not think it is worth the effort.

Reply:
> How can you get that precision?

@lpinca this is the code I use for loop benchmark:
```javascript
const iterations = 1e9
const cycleNS = null

const fn=()=>{
   // code to benchmark
}

let i=iterations,start=process.hrtime.bigint()
while(i--)fn()
const ns=Number(process.hrtime.bigint()-start)/iterations
console.log(ns.toFixed(2),'ns')
cycleNS&&console.log((ns/cycleNS).toFixed(2),'cycles')
```
You want to adjust the iteration number `1e9` so it runs for at least a second so the JS engine optimizes the function. You can count how many cpu cycles a function takes, the nanoseconds fall on clear intervals of the cpu cycle time. If you bench an empty function that should take 2 cycles, divide the result by 2 for your cycle time, for me 1 cycle is .228 nanoseconds, which happens to match my CPU speed 4.7GHz = 4.7 cycles per nanosecond. Change `cycleNS` to your cycle time to add # cycles to the result log

Using this you can see that Uint32Array can be faster to work with than Buffer, here are some cpu cycles per expression:
```javascript
uint32[0] = 123 // 2 cycles
buf.writeUInt32BE(123) // 11 cycles
buf[0] = 1; buf[1] = 2; buf[2] = 3; buf[3] = 4 // 5 cycles

uint32[0] // 2 cycles
buf.readUInt32BE() // 8 cycles
buf[0]; buf[1]; buf[2]; buf[3] // 2 cycles

Math.random() // 30 cycles
crypto.randomFillSync(buf) // 6200 cycles
```
These numbers are so small they don't really matter but it is interesting to see how it works